pkg/restore/cdclog/buffer.go

```diff
@@ -0,0 +1,174 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cdclog
+
+import (
+	"context"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/types"
+	"go.uber.org/zap"
+)
+
+// TableBuffer represents the kv buffer of this table.
+// we restore one tableBuffer in one goroutine.
+// this is the concurrent unit of log restore.
+type TableBuffer struct {
+	KvPairs []Row
+	Size    int64
+
+	KvEncoder Encoder
+	tableInfo table.Table
+
+	flushKVPairs int
+
+	colNames []string
+	colPerm  []int
+}
+
+// NewTableBuffer creates TableBuffer.
+func NewTableBuffer(tbl table.Table, flushKVPairs int) *TableBuffer {
+	kvEncoder := NewTableKVEncoder(tbl, &SessionOptions{
+		Timestamp: time.Now().Unix(),
+		// TODO make it config
+		RowFormatVersion: "1",
+	})
+
+	tb := &TableBuffer{
+		KvPairs:      make([]Row, 0, flushKVPairs),
+		KvEncoder:    kvEncoder,
+		flushKVPairs: flushKVPairs,
+	}
+	tb.ReloadMeta(tbl)
+	return tb
+}
+
+// TableID return this table id.
+func (t *TableBuffer) TableID() int64 {
+	return t.tableInfo.Meta().ID
+}
+
+// ReloadMeta reload columns after
+// 1. table buffer created.
+// 2. every ddl executed.
+func (t *TableBuffer) ReloadMeta(tbl table.Table) {
+	columns := tbl.Meta().Cols()
+	colNames := make([]string, 0, len(columns))
+	colPerm := make([]int, 0, len(columns)+1)
+
+	for i, col := range columns {
+		colNames = append(colNames, col.Name.String())
+		colPerm = append(colPerm, i)
+	}
+	if TableHasAutoRowID(tbl.Meta()) {
+		colPerm = append(colPerm, -1)
+	}
+	t.tableInfo = tbl
+	t.colNames = colNames
+	t.colPerm = colPerm
+}
+
+func (t *TableBuffer) translateToDatum(row map[string]column) ([]types.Datum, error) {
+	cols := make([]types.Datum, 0, len(row))
+	for _, col := range t.colNames {
+		val, err := row[col].toDatum()
+		if err != nil {
+			return nil, err
+		}
+		log.Debug("translate to datum", zap.String("col", col), zap.Stringer("val", val))
+		cols = append(cols, val)
+	}
+	return cols, nil
+}
+
+func (t *TableBuffer) appendRow(
+	row map[string]column,
+	item *SortItem,
+	encodeFn func(row []types.Datum,
+		rowID int64,
+		columnPermutation []int) (Row, error),
+) error {
+	cols, err := t.translateToDatum(row)
+	if err != nil {
+		return err
+	}
+	pair, err := encodeFn(cols, item.RowID, t.colPerm)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	t.KvPairs = append(t.KvPairs, pair)
+	t.Size++
+	return nil
+}
+
+// Append appends the item to this buffer.
+func (t *TableBuffer) Append(ctx context.Context, item *SortItem) error {
+	log.Debug("Append item to buffer",
+		zap.Stringer("table", t.tableInfo.Meta().Name),
+	)
+	row := item.Meta.(*MessageRow)
+
+	select {
+	case <-ctx.Done():
```

Comment:
why need to do a select here, this append should be quite fast, are there any blocks? the `Context` parameter is only used in the select check, I think this isn't meaningful

Reply:
context has been removed.